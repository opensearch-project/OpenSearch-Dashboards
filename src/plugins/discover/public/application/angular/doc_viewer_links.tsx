/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { DocViewerLinks } from '../components/doc_viewer_links/doc_viewer_links';

export function createDocViewerLinksDirective(reactDirective: any) {
  return reactDirective(
    (props: any) => {
      return <DocViewerLinks {...props} />;
    },
    [
      'hit',
      ['indexPattern', { watchDepth: 'reference' }],
      ['columns', { watchDepth: 'collection' }],
    ],
    {
      restrict: 'E',
      scope: {
        hit: '=',
        indexPattern: '=',
        columns: '=?',
      },
    }
  );
}
