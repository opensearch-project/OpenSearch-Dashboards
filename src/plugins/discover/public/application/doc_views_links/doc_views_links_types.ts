/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiListGroupItemProps } from '@elastic/eui';
import { OpenSearchSearchHit } from '../doc_views/doc_views_types';
import { IndexPattern } from '../../../../data/public';

export interface DocViewLink extends EuiListGroupItemProps {
  href?: string;
  order: number;
  generateCb?(
    renderProps: any
  ): {
    url: string;
    hide?: boolean;
  };
}

export interface DocViewLinkRenderProps {
  columns?: string[];
  hit: OpenSearchSearchHit;
  indexPattern: IndexPattern;
}
