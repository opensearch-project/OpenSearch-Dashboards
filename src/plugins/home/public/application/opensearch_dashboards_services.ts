/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  ChromeStart,
  DocLinksStart,
  HttpStart,
  NotificationsSetup,
  OverlayStart,
  SavedObjectsClientContract,
  IUiSettingsClient,
  ApplicationStart,
} from 'opensearch-dashboards/public';
import { UiStatsMetricType } from '@osd/analytics';
import { TelemetryPluginStart } from '../../../telemetry/public';
import { UrlForwardingStart } from '../../../url_forwarding/public';
import { TutorialService } from '../services/tutorials';
import { FeatureCatalogueRegistry } from '../services/feature_catalogue';
import { EnvironmentService } from '../services/environment';
import { ConfigSchema } from '../../config';
import { HomePluginBranding } from '..';
import { DataSourcePluginStart } from '../../../data_source/public';

export interface HomeOpenSearchDashboardsServices {
  indexPatternService: any;
  opensearchDashboardsVersion: string;
  chrome: ChromeStart;
  application: ApplicationStart;
  uiSettings: IUiSettingsClient;
  urlForwarding: UrlForwardingStart;
  homeConfig: ConfigSchema;
  featureCatalogue: FeatureCatalogueRegistry;
  http: HttpStart;
  savedObjectsClient: SavedObjectsClientContract;
  toastNotifications: NotificationsSetup['toasts'];
  banners: OverlayStart['banners'];
  trackUiMetric: (type: UiStatsMetricType, eventNames: string | string[], count?: number) => void;
  getBasePath: () => string;
  docLinks: DocLinksStart;
  addBasePath: (url: string) => string;
  environmentService: EnvironmentService;
  telemetry?: TelemetryPluginStart;
  tutorialService: TutorialService;
  injectedMetadata: {
    getInjectedVar: (name: string, defaultValue?: any) => unknown;
    getBranding: () => HomePluginBranding;
  };
  dataSource?: DataSourcePluginStart;
}

let services: HomeOpenSearchDashboardsServices | null = null;

export function setServices(newServices: HomeOpenSearchDashboardsServices) {
  services = newServices;
}

export function getServices() {
  if (!services) {
    throw new Error(
      'OpenSearch Dashboards services not set - are you trying to import this module from outside of the home app?'
    );
  }
  return services;
}
