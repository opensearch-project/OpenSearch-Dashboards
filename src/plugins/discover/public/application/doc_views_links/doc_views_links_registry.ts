/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DocViewLink } from './doc_views_links_types';

export class DocViewsLinksRegistry {
  private docViewsLinks: DocViewLink[] = [];

  addDocViewLink(docViewLink: DocViewLink) {
    this.docViewsLinks.push(docViewLink);
  }

  getDocViewsLinksSorted() {
    return this.docViewsLinks.sort((a, b) => (Number(a.order) > Number(b.order) ? 1 : -1));
  }
}
