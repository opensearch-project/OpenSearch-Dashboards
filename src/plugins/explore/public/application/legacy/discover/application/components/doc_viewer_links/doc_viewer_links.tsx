/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiFlexGroup, EuiFlexItem, EuiListGroupItemProps, EuiLink } from '@elastic/eui';
import { getDocViewsLinksRegistry } from '../../../opensearch_dashboards_services';
import { DocViewLinkRenderProps } from '../../doc_views_links/doc_views_links_types';

export function DocViewerLinks(renderProps: DocViewLinkRenderProps) {
  const listItems = getDocViewsLinksRegistry()
    .getDocViewsLinksSorted()
    .filter((item) => !(item.generateCb && item.generateCb(renderProps)?.hide))
    .map((item) => {
      const { generateCb, href, ...props } = item;
      const listItem: EuiListGroupItemProps = {
        'data-test-subj': `docTableRowAction`,
        ...props,
        href: generateCb ? generateCb(renderProps).url : href,
      };

      return listItem;
    });

  return (
    <EuiFlexGroup gutterSize="m" justifyContent="flexEnd">
      {listItems.map((item, index) => (
        <EuiFlexItem key={index} grow={false}>
          <EuiLink
            href={item.href}
            target="_blank"
            style={{ fontWeight: 'normal' }}
            data-test-subj={`${item['data-test-subj']}-${index}`}
          >
            {item.label}
          </EuiLink>
        </EuiFlexItem>
      ))}
    </EuiFlexGroup>
  );
}
