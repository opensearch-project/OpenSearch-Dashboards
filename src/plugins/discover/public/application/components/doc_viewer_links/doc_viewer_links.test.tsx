/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { shallow } from 'enzyme';
import { DocViewerLinks } from './doc_viewer_links';
import { getDocViewsLinksRegistry } from '../../../opensearch_dashboards_services';
import { DocViewLinkRenderProps } from '../doc_views/doc_views_links/doc_views_links_types';

jest.mock('../../../opensearch_dashboards_services', () => {
  let registry: any[] = [];
  return {
    getDocViewsLinksRegistry: () => ({
      addDocViewLink(view: any) {
        registry.push(view);
      },
      getDocViewsLinksSorted() {
        return registry;
      },
      resetRegistry: () => {
        registry = [];
      },
    }),
  };
});

beforeEach(() => {
  (getDocViewsLinksRegistry() as any).resetRegistry();
  jest.clearAllMocks();
});

test('Render <DocViewerLink/> with 2 different links', () => {
  const registry = getDocViewsLinksRegistry();
  registry.addDocViewLink({
    order: 10,
    label: 'generateCb link',
    generateCb: () => ({
      url: 'aaa',
    }),
  });
  registry.addDocViewLink({ order: 20, label: 'href link', href: 'bbb' });

  const renderProps = { hit: {} } as DocViewLinkRenderProps;

  const wrapper = shallow(<DocViewerLinks {...renderProps} />);

  expect(wrapper).toMatchSnapshot();
});

test('Dont Render <DocViewerLink/> if generateCb.hide', () => {
  const registry = getDocViewsLinksRegistry();
  registry.addDocViewLink({
    order: 10,
    label: 'generateCb link',
    generateCb: () => ({
      url: 'aaa',
      hide: true,
    }),
  });

  const renderProps = { hit: {} } as DocViewLinkRenderProps;

  const wrapper = shallow(<DocViewerLinks {...renderProps} />);

  expect(wrapper).toMatchSnapshot();
});
