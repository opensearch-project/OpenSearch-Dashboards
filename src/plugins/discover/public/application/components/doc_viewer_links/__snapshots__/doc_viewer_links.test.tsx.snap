// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dont Render <DocViewerLink/> if generateCb.hide 1`] = `
<EuiFlexGroup
  gutterSize="m"
  justifyContent="flexEnd"
/>
`;

exports[`Render <DocViewerLink/> with 2 different links 1`] = `
<EuiFlexGroup
  gutterSize="m"
  justifyContent="flexEnd"
>
  <EuiFlexItem
    grow={false}
    key="0"
  >
    <EuiLink
      data-test-subj="docTableRowAction-0"
      href="aaa"
      style={
        Object {
          "fontWeight": "normal",
        }
      }
      target="_blank"
    >
      generateCb link
    </EuiLink>
  </EuiFlexItem>
  <EuiFlexItem
    grow={false}
    key="1"
  >
    <EuiLink
      data-test-subj="docTableRowAction-1"
      href="bbb"
      style={
        Object {
          "fontWeight": "normal",
        }
      }
      target="_blank"
    >
      href link
    </EuiLink>
  </EuiFlexItem>
</EuiFlexGroup>
`;
